From 97df39cf8ceac9a8c20604e4cf6920039c9ad8a2 Mon Sep 17 00:00:00 2001
From: Uraj Sasan <usasan@qti.qualcomm.com>
Date: Tue, 12 Aug 2025 14:04:17 +0530
Subject: [PATCH] QSDK:PLATFORM:wifi: ath12k: INI files location changes

Update the INI file access paths for ath12k to use /ini and /ini/internal
instead of the previous /lib/fw/ath12k location. This change enables
centralized storage and better organization of configuration files

Signed-off-by: Uraj Sasan <usasan@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/core.c      | 12 ++--
 drivers/net/wireless/ath/ath12k/ini.c       | 70 +++++++++++++--------
 drivers/net/wireless/ath/ath12k/ini.h       |  1 -
 drivers/net/wireless/ath/ath12k/ini_parse.c | 10 ++-
 4 files changed, 59 insertions(+), 34 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index 458820b..3b2bb32 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -1051,6 +1051,7 @@ static void ath12k_core_stop(struct ath12k_base *ab)
 	ath12k_hif_stop(ab);
 	ath12k_wmi_detach(ab);
 	ath12k_dp_cmn_device_deinit(ab->dp);
+	ath12k_cfg_deinit(ab);
 
 	/* De-Init of components as needed */
 }
@@ -1191,7 +1192,6 @@ static void ath12k_core_soc_destroy(struct ath12k_base *ab)
 	ath12k_reg_free(ab);
 	ath12k_debugfs_soc_destroy(ab);
 	ath12k_qmi_deinit_service(ab);
-	ath12k_cfg_deinit(ab);
 }
 
 static int ath12k_core_mlo_shmem_per_device_crash_info_addresses(
@@ -1743,11 +1743,6 @@ core_pdev_create:
 		mutex_unlock(&ab->core_lock);
 	}
 
-	if (ath12k_cfg_init(ab))
-		ath12k_err(ab, "failed to initialize INI data in driver\n");
-	else
-		ath12k_err(ab, "initialize INI data in driver\n");
-
 	return 0;
 
 err:
@@ -1909,6 +1904,11 @@ int ath12k_core_qmi_firmware_ready(struct ath12k_base *ab)
 		goto err_firmware_stop;
 	}
 
+	if (ath12k_cfg_init(ab))
+		ath12k_err(ab, "failed to initialize INI data in driver\n");
+	else
+		ath12k_info(ab, "initialized INI data in driver\n");
+
 	mutex_lock(&ag->mutex);
 	mutex_lock(&ab->core_lock);
 
diff --git a/drivers/net/wireless/ath/ath12k/ini.c b/drivers/net/wireless/ath/ath12k/ini.c
index 947e93d..87c6cac 100644
--- a/drivers/net/wireless/ath/ath12k/ini.c
+++ b/drivers/net/wireless/ath/ath12k/ini.c
@@ -98,8 +98,7 @@ ath12k_cfg_int_item_handler(struct ath12k_cfg_value_store *store,
 	s32 *store_value = ath12k_cfg_value_ptr(store, meta);
 	s32 value;
 
-	ret = kstrtoint(str_value, 10, &value);
-
+	ret = kstrtoint(str_value, 0, &value);
 	if (ret) {
 		ath12k_err(NULL,
 			   "%s=%s -Invalid format (status %d) Using default %d",
@@ -167,10 +166,10 @@ ath12k_cfg_uint_item_handler(struct ath12k_cfg_value_store *store,
 	min = (uint32_t)meta->min;
 	max = (uint32_t)meta->max;
 
-	ret = kstrtouint(str_value, 10, &value);
+	ret = kstrtouint(str_value, 0, &value);
 	if (ret) {
 		ath12k_err(NULL,
-			   "%s=%s-Invalid format (status %d); Using default %u",
+			   "%s = %s - Invalid format (status %d) using default %d",
 			   meta->name, str_value, ret, *store_value);
 		return;
 	}
@@ -596,6 +595,7 @@ struct ath12k_cfg_values *ath12k_cfg_get_values(struct ath12k_base *ab)
 {
 	return &ath12k_cfg_get_ctx(ab)->store->values;
 }
+EXPORT_SYMBOL(ath12k_cfg_get_values);
 
 /**
  * ath12k_cfg_ini_parse_to_store - Parse the ini file to populate the cfg store
@@ -610,6 +610,7 @@ ath12k_cfg_ini_parse_to_store(const char *path,
 	int ret;
 
 	ret = ath12k_ini_parse(path, store, ath12k_cfg_ini_item_handler);
+
 	if (ret)
 		ath12k_err(NULL,
 			   "Failed to parse *.ini file @ %s; status:%d",
@@ -935,7 +936,7 @@ ath12k_cfg_ini_config_print(struct ath12k_base *ab, uint8_t *buf,
 int ath12k_cfg_dispatcher_init(struct ath12k_base *ab)
 {
 	if (ath12k_cfg_is_init) {
-		ath12k_err(ab, "cfg dispatcher already initialized\n");
+		ath12k_dbg(ab, ATH12K_DBG_INI, "cfg dispatcher already initialized\n");
 		return 0;
 	}
 
@@ -949,7 +950,7 @@ int ath12k_cfg_dispatcher_init(struct ath12k_base *ab)
 int ath12k_cfg_dispatcher_deinit(struct ath12k_base *ab)
 {
 	if (!ath12k_cfg_is_init) {
-		ath12k_warn(ab, "cfg dispatcher already de-initialized\n");
+		ath12k_dbg(ab, ATH12K_DBG_INI, "cfg dispatcher already de-initialized\n");
 		return -EINVAL;
 	}
 
@@ -965,9 +966,22 @@ int ath12k_cfg_get_ini_file_name(u32 target_type, struct ath12k_ini_file *ini)
 	int ret = 0;
 
 	switch (target_type) {
-	case TARGET_TYPE_QCN9224:
-		ini->external = "QCN9224.ini";
-		ini->internal = "QCN9224_i.ini";
+	case ATH12K_HW_QCN9274_HW10:
+	case ATH12K_HW_QCN9274_HW20:
+		ini->external = "QCN9274.ini";
+		ini->internal = "QCN9274_i.ini";
+		break;
+	case ATH12K_HW_IPQ5332_HW10:
+		ini->external = "IPQ5332.ini";
+		ini->internal = "IPQ5332_i.ini";
+		break;
+	case ATH12K_HW_IPQ5424_HW10:
+		ini->external = "IPQ5424.ini";
+		ini->internal = "IPQ5424_i.ini";
+		break;
+	case ATH12K_HW_QCN6432_HW10:
+		ini->external = "QCN6432.ini";
+		ini->internal = "QCN6432_i.ini";
 		break;
 	default:
 		ini->external = NULL;
@@ -985,36 +999,42 @@ int ath12k_cfg_init(struct ath12k_base *ab)
 {
 	struct ath12k_ini_file ini;
 	char ini_buf[ATH12K_CFG_FILE_NAME_MAX];
-	char ini_buf_ext[ATH12K_CFG_FILE_NAME_MAX];
 	char global_file[ATH12K_CFG_FILE_NAME_MAX];
+	char global_file_i[ATH12K_CFG_FILE_NAME_MAX];
 	int ret;
 
 	ret = ath12k_cfg_dispatcher_init(ab);
-	if (ret)
-		ath12k_warn(ab, "cfg dispatcher already initialized\n");
-
-	scnprintf(global_file, sizeof(global_file), "%s/global.ini",
-		  ATH12K_FW_DIR);
+	if (ret) {
+		ath12k_err(ab, "Failed to initialize cfg dispatcher\n");
+		return ret;
+	}
+	scnprintf(global_file, sizeof(global_file), "global.ini");
 	if (ath12k_cfg_parse(global_file)) {
-		ath12k_err(ab, "Failed to parse the global ini file\n");
+		ath12k_err(ab, "Failed to parse the global ini %s\n", global_file);
+		ath12k_cfg_deinit(ab);
+		return -EINVAL;
+	}
+
+	scnprintf(global_file_i, sizeof(global_file_i), "internal/global_i.ini");
+	if (ath12k_cfg_parse(global_file_i)) {
+		ath12k_err(ab, "Failed to parse global_i ini %s\n", global_file_i);
+		ath12k_cfg_deinit(ab);
 		return -EINVAL;
 	}
-	scnprintf(ini_buf, sizeof(ini_buf), "%s/internal/", ATH12K_FW_DIR);
-	scnprintf(ini_buf_ext, sizeof(ini_buf_ext), "%s/", ATH12K_FW_DIR);
+
+	scnprintf(ini_buf, sizeof(ini_buf), "internal/");
 	if (ath12k_cfg_on_create(ab)) {
 		ath12k_err(ab, "Failed to create the cfg store context \n");
+		ath12k_cfg_deinit(ab);
 		return -EINVAL;
 	}
 	/* Parse the target specific INI */
-	ret = ath12k_cfg_get_ini_file_name(TARGET_TYPE_QCN9224, &ini);
+	ret = ath12k_cfg_get_ini_file_name(ab->hw_rev, &ini);
 
 	if (!ret) {
 		strlcat(ini_buf, ini.internal,
 			strlen(ini.internal) + strlen(ini_buf) + 1);
-		strlcat(ini_buf_ext,
-			ini.external,
-			strlen(ini.external) + strlen(ini_buf_ext) + 1);
-		if (ath12k_cfg_ab_parse(ab, ini_buf_ext) == 0)
+		if (ath12k_cfg_ab_parse(ab, ini.external) == 0)
 			ath12k_cfg_parse_to_store(ab, ini_buf);
 		else
 			ath12k_cfg_ab_parse(ab, ini_buf);
@@ -1040,11 +1060,11 @@ void ath12k_cfg_parse_pdev_section(struct ath12k_base *ab)
 	int ret;
 	u32 i;
 
-	scnprintf(ini_buf, sizeof(ini_buf), "%s/internal/", ATH12K_FW_DIR);
+	scnprintf(ini_buf, sizeof(ini_buf), "internal/");
 	if (!ab->cfg_ctx)
 		return;
 	/* Parse the target specific INI */
-	ret = ath12k_cfg_get_ini_file_name(TARGET_TYPE_QCN9224, &ini);
+	ret = ath12k_cfg_get_ini_file_name(ab->hw_rev, &ini);
 	if (ret)
 		return;
 
diff --git a/drivers/net/wireless/ath/ath12k/ini.h b/drivers/net/wireless/ath/ath12k/ini.h
index 6beabf8..b7ec5e9 100644
--- a/drivers/net/wireless/ath/ath12k/ini.h
+++ b/drivers/net/wireless/ath/ath12k/ini.h
@@ -74,7 +74,6 @@
 #define ATH12K_CFG_META_NAME_LENGTH_MAX 256
 #define ATH12K_CFG_INI_LENGTH_MAX 128
 #define ATH12K_CFG_FILE_NAME_MAX 128
-#define TARGET_TYPE_QCN9224      32
 #define ATH12K_MAX_CFG 0xffffffff
 
 #define ATH12K_CFG_ALL \
diff --git a/drivers/net/wireless/ath/ath12k/ini_parse.c b/drivers/net/wireless/ath/ath12k/ini_parse.c
index e71fbf8..85c890d 100644
--- a/drivers/net/wireless/ath/ath12k/ini_parse.c
+++ b/drivers/net/wireless/ath/ath12k/ini_parse.c
@@ -21,6 +21,11 @@ const struct firmware *ini_file_read(const char *path)
 		ath12k_dbg(NULL, ATH12K_DBG_INI, "Failed to read file %s", path);
 		return ERR_PTR(ret);
 	}
+	if (fw && !fw->size)
+		return ERR_PTR(-ENOENT);
+
+	ath12k_dbg(NULL, ATH12K_DBG_INI, "boot firmware request %s size %zu\n",
+		   path, fw->size);
 
 	return fw;
 }
@@ -142,9 +147,10 @@ int ath12k_ini_parse(const char *ini_path, void *context,
 	char *end;
 
 	fw = ini_file_read(ini_path);
+
 	if (IS_ERR(fw)) {
 		ath12k_dbg(NULL, ATH12K_DBG_INI, "Failed to read *.ini file @ %s", ini_path);
-		return ret;
+		return PTR_ERR(fw);
 	}
 
 	/* foreach line */
@@ -203,7 +209,7 @@ int ath12k_ini_section_parse(const char *ini_path, void *context,
 	fw = ini_file_read(ini_path);
 	if (IS_ERR(fw)) {
 		ath12k_dbg(NULL, ATH12K_DBG_INI, "Failed to read *.ini file @ %s", ini_path);
-		return ret;
+		return PTR_ERR(fw);
 	}
 
 	/* foreach line */
-- 
2.34.1

